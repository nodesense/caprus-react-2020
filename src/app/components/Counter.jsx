// Counter.jsx
import React from 'react';
import PropTypes from 'prop-types';

// class component
// class state - only available for class component

// STATE: A Component data, which can be mutated/changed
// props are read only, state can be used for get and set

class Counter extends React.Component {
    // ES.next
    static propTypes = {
        startValue: PropTypes.number.isRequired
    }

    static defaultProps = {
        startValue: 0
    }

    constructor(props) {
        super(props) // mandatory

        // initialize state 
        this.state = {
            counter : props.startValue
        }
    }

    // react shall pass event object as its first argument
    increment(e) {
        console.log('increment called', e);
        console.log('this value is ', this)
    }

    // keyword
    // to create v.doms and return v.doms
    render() {
        // this.props contains props value
        console.log("Counter render props ", this.props);
        const {startValue} = this.props;

        return (  
            <div>
                <h2>Counter</h2>
                <p>Start Value {startValue}</p>

                <p>Counter value {this.state.counter}</p>
                {/* react events take callback as reference */}
                <button onClick={this.increment}>+1</button>
            </div>
        )
    }
}

export default Counter;