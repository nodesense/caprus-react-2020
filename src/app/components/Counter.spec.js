import React from 'react';
import renderer from 'react-test-renderer';

import Counter from './Counter';


test('Link changes the class when hovered', () => {
    const component = renderer.create(
      <Counter startValue={100} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component.getInstance().increment();
   // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot(); // 101
});

// npm test
