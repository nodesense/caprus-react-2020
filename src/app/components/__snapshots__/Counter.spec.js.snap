// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<div>
  <h2>
    Counter
  </h2>
  <p>
    Start Value 
    100
  </p>
  <p>
    Counter value 
    100
  </p>
  <button
    onClick={[Function]}
  >
    Set State
  </button>
  <button
    onClick={[Function]}
  >
    forceUpdate
  </button>
  <button
    onClick={[Function]}
  >
    +1 CRASH due this
  </button>
  <button
    onClick={[Function]}
  >
    incrementTwiceBug
  </button>
  <button
    onClick={[Function]}
  >
    incrementTwice Render twice Inefficient
  </button>
  <button
    onClick={[Function]}
  >
    decrementTwice
  </button>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <div
    onClick={[Function]}
  >
    <button
      onClick={[Function]}
    >
      -1
    </button>
  </div>
</div>
`;

exports[`Link changes the class when hovered 2`] = `
<div>
  <h2>
    Counter
  </h2>
  <p>
    Start Value 
    100
  </p>
  <p>
    Counter value 
    101
  </p>
  <button
    onClick={[Function]}
  >
    Set State
  </button>
  <button
    onClick={[Function]}
  >
    forceUpdate
  </button>
  <button
    onClick={[Function]}
  >
    +1 CRASH due this
  </button>
  <button
    onClick={[Function]}
  >
    incrementTwiceBug
  </button>
  <button
    onClick={[Function]}
  >
    incrementTwice Render twice Inefficient
  </button>
  <button
    onClick={[Function]}
  >
    decrementTwice
  </button>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <div
    onClick={[Function]}
  >
    <button
      onClick={[Function]}
    >
      -1
    </button>
  </div>
</div>
`;
